packages/firestore/src/core/sync_engine.ts

```diff
@@ -602,6 +609,62 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
     }
   }
 
+  /**
+   * Takes a snapshot of current mutation queue, and register a user callback
+   * which will be called when all those mutations in the snapshot are either
+   * accepted or rejected by the server.
+   */
+  async registerPendingWritesCallback(callback: Deferred<void>): Promise<void> {
+    if (!this.remoteStore.canUseNetwork()) {
+      log.debug(
+        LOG_TAG,
+        'The network is disabled. The task returned by ' +
+          "'awaitPendingWrites()' will not complete until the network is enabled."
+      );
+    }
+
+    const largestBatchId = await this.localStore.getHighestUnacknowledgedBatchId();
+
+    if (largestBatchId === BATCHID_UNKNOWN) {
+      // Trigger the callback right away if there is no pending writes at the moment.
+      callback.resolve();
+      return Promise.resolve();
+    }
+
+    const callbacks = this.pendingWritesCallbacks.get(largestBatchId) || [];
```

Comment:
It would be beneficial to verify what the transpiled JS looks like, but this could potentially be:

```
const callbacks = [... this.pendingWritesCallbacks.get(largestBatchId), callback]
```

If we don't transpile the spread operator away, there is some chance we are breaking users, so this might be not worth exploring.

Reply:
You can invoke `yarn build`  in "packages/firestore". It will spit out a couple different files into "packages/firestore/dist"

In packages/firestore/dist/index.esm.js, your current code looks as such:

```
callbacks = this.pendingWritesCallbacks.get(highestBatchId) || [];
callbacks.push(callback);
this.pendingWritesCallbacks.set(highestBatchId, callbacks);
```

It looks like we have to specify different compiler settings to make my suggestion work (--downlevelIterators)/ Since it is just an optional suggestion, let's keep it as is.